File: src/types/modules.ts
/**
 * TypeScript types for Module Registry and Hub Type System
 * Controls which features/modules are available per hub type
 */

// ============================================================
// HUB TYPES
// ============================================================

/**
 * Hub types determine what kind of functionality a data table supports
 * - activities: Event/activity management (supports Pulse scanning)
 * - applications: Application review workflows (supports Review Workflows)
 * - data: General data management (basic tables)
 */
export type HubType = 'activities' | 'applications' | 'data';

export const HUB_TYPE_LABELS: Record<HubType, string> = {
  activities: 'Activities Hub',
  applications: 'Applications Hub',
  data: 'Data Hub',
};

export const HUB_TYPE_DESCRIPTIONS: Record<HubType, string> = {
  activities: 'For events, activities, and attendance tracking with Pulse scanning support',
  applications: 'For application review workflows with multi-stage review and rubrics',
  data: 'For general data management with Airtable-like tables',
};

export const HUB_TYPE_ICONS: Record<HubType, string> = {
  activities: 'calendar-check',
  applications: 'file-text',
  data: 'table',
};

export const HUB_TYPE_COLORS: Record<HubType, string> = {
  activities: '#10B981', // Emerald
  applications: '#6366F1', // Indigo
  data: '#3B82F6', // Blue
};

// ============================================================
// MODULE DEFINITIONS
// ============================================================

export type ModuleCategory = 'core' | 'productivity' | 'communication' | 'integration';

/**
 * Module Definition - describes a feature that can be enabled on a hub
 */
export interface ModuleDefinition {
  id: string;
  name: string;
  description?: string;
  icon?: string;
  category: ModuleCategory;
  is_premium: boolean;
  is_beta: boolean;
  is_deprecated: boolean;
  available_for_hub_types: HubType[];
  dependencies: string[];
  settings_schema?: Record<string, any>;
  display_order: number;
  created_at?: string;
  updated_at?: string;
}

/**
 * Known module IDs for type safety
 */
export type KnownModuleId =
  | 'tables'
  | 'views'
  | 'forms'
  | 'pulse'
  | 'attendance'
  | 'calendar'
  | 'review_workflow'
  | 'rubrics'
  | 'reviewer_portal'
  | 'decision_logic'
  | 'analytics'
  | 'export'
  | 'notifications'
  | 'email_templates';

// ============================================================
// HUB MODULE CONFIGURATION
// ============================================================

/**
 * Hub Module Config - per-table module enablement and settings
 */
export interface HubModuleConfig {
  id: string;
  table_id: string;
  module_id: string;
  is_enabled: boolean;
  settings?: Record<string, any>;
  enabled_by?: string;
  enabled_at?: string;
  updated_at?: string;
}

/**
 * Module with its enabled status for a specific hub
 */
export interface ModuleWithStatus extends ModuleDefinition {
  is_enabled: boolean;
  settings?: Record<string, any>;
}

/**
 * API response for hub modules
 */
export interface HubModulesResponse {
  table_id: string;
  hub_type: HubType;
  enabled_modules: ModuleWithStatus[];
  available_modules: ModuleWithStatus[];
}

// ============================================================
// REQUEST/RESPONSE TYPES
// ============================================================

export interface EnableModuleRequest {
  module_id: string;
  settings?: Record<string, any>;
}

export interface UpdateHubTypeRequest {
  hub_type: HubType;
}

export interface CreateHubRequest {
  workspace_id: string;
  name: string;
  slug?: string;
  description?: string;
  hub_type: HubType;
  icon?: string;
  color?: string;
}

// ============================================================
// STATIC MODULE REGISTRY (Frontend Mirror)
// ============================================================

/**
 * Static registry of all modules - mirrors database module_definitions
 * Use this for UI rendering when database isn't available
 */
export const MODULE_REGISTRY: Record<KnownModuleId, Omit<ModuleDefinition, 'created_at' | 'updated_at'>> = {
  tables: {
    id: 'tables',
    name: 'Tables',
    description: 'Airtable-like data tables with columns and rows',
    icon: 'table',
    category: 'core',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['activities', 'applications', 'data'],
    dependencies: [],
    display_order: 1,
  },
  views: {
    id: 'views',
    name: 'Views',
    description: 'Grid, Kanban, Calendar, and Gallery views',
    icon: 'layout',
    category: 'core',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['activities', 'applications', 'data'],
    dependencies: ['tables'],
    display_order: 2,
  },
  forms: {
    id: 'forms',
    name: 'Forms',
    description: 'Create intake forms that populate tables',
    icon: 'file-text',
    category: 'core',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['activities', 'applications', 'data'],
    dependencies: ['tables'],
    display_order: 3,
  },
  pulse: {
    id: 'pulse',
    name: 'Pulse Scanning',
    description: 'Barcode/QR code check-in and attendance tracking',
    icon: 'scan-line',
    category: 'productivity',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['activities'],
    dependencies: ['tables'],
    display_order: 10,
  },
  attendance: {
    id: 'attendance',
    name: 'Attendance Tracking',
    description: 'Track attendance for activities and events',
    icon: 'user-check',
    category: 'productivity',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['activities'],
    dependencies: ['tables', 'pulse'],
    display_order: 11,
  },
  calendar: {
    id: 'calendar',
    name: 'Calendar Integration',
    description: 'Sync activities with calendar apps',
    icon: 'calendar',
    category: 'integration',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['activities'],
    dependencies: ['tables'],
    display_order: 12,
  },
  review_workflow: {
    id: 'review_workflow',
    name: 'Review Workflows',
    description: 'Multi-stage application review with stages and reviewers',
    icon: 'workflow',
    category: 'productivity',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['applications'],
    dependencies: ['tables', 'forms'],
    display_order: 20,
  },
  rubrics: {
    id: 'rubrics',
    name: 'Scoring Rubrics',
    description: 'Create rubrics for consistent scoring',
    icon: 'list-checks',
    category: 'productivity',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['applications'],
    dependencies: ['review_workflow'],
    display_order: 21,
  },
  reviewer_portal: {
    id: 'reviewer_portal',
    name: 'External Reviewer Portal',
    description: 'Allow external reviewers to score applications',
    icon: 'external-link',
    category: 'productivity',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['applications'],
    dependencies: ['review_workflow', 'rubrics'],
    display_order: 22,
  },
  decision_logic: {
    id: 'decision_logic',
    name: 'Decision Logic',
    description: 'Automate advancement and rejection based on scores',
    icon: 'git-branch',
    category: 'productivity',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['applications'],
    dependencies: ['review_workflow'],
    display_order: 23,
  },
  analytics: {
    id: 'analytics',
    name: 'Advanced Analytics',
    description: 'Charts, reports, and dashboards',
    icon: 'chart-bar',
    category: 'productivity',
    is_premium: true,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['activities', 'applications', 'data'],
    dependencies: ['tables'],
    display_order: 30,
  },
  export: {
    id: 'export',
    name: 'Advanced Export',
    description: 'Export to Excel, PDF, and custom formats',
    icon: 'download',
    category: 'productivity',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['activities', 'applications', 'data'],
    dependencies: ['tables'],
    display_order: 31,
  },
  notifications: {
    id: 'notifications',
    name: 'Notifications',
    description: 'Email and in-app notifications',
    icon: 'bell',
    category: 'communication',
    is_premium: false,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['activities', 'applications', 'data'],
    dependencies: [],
    display_order: 40,
  },
  email_templates: {
    id: 'email_templates',
    name: 'Email Templates',
    description: 'Create reusable email templates',
    icon: 'mail',
    category: 'communication',
    is_premium: true,
    is_beta: false,
    is_deprecated: false,
    available_for_hub_types: ['applications'],
    dependencies: ['notifications'],
    display_order: 41,
  },
};

// ============================================================
// HELPER FUNCTIONS
// ============================================================

/**
 * Get all modules available for a specific hub type
 */
export function getModulesForHubType(hubType: HubType): ModuleDefinition[] {
  return Object.values(MODULE_REGISTRY)
    .filter(module => module.available_for_hub_types.includes(hubType))
    .sort((a, b) => a.display_order - b.display_order) as ModuleDefinition[];
}

/**
 * Check if a module can be enabled (all dependencies are met)
 */
export function canEnableModule(
  moduleId: KnownModuleId,
  enabledModuleIds: string[]
): boolean {
  const module = MODULE_REGISTRY[moduleId];
  if (!module) return false;
  
  return module.dependencies.every(dep => enabledModuleIds.includes(dep));
}

/**
 * Get missing dependencies for a module
 */
export function getMissingDependencies(
  moduleId: KnownModuleId,
  enabledModuleIds: string[]
): string[] {
  const module = MODULE_REGISTRY[moduleId];
  if (!module) return [];
  
  return module.dependencies.filter(dep => !enabledModuleIds.includes(dep));
}

/**
 * Get default modules for a hub type (auto-enabled on creation)
 */
export function getDefaultModulesForHubType(hubType: HubType): KnownModuleId[] {
  const defaults: Record<HubType, KnownModuleId[]> = {
    activities: ['tables', 'views', 'pulse', 'attendance'],
    applications: ['tables', 'views', 'forms', 'review_workflow', 'rubrics'],
    data: ['tables', 'views'],
  };
  return defaults[hubType];
}

File: go-backend/models/modules.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// HubType enum values
const (
	HubTypeActivities   = "activities"
	HubTypeApplications = "applications"
	HubTypeData         = "data"
)

// ModuleCategory enum values
const (
	ModuleCategoryCore          = "core"
	ModuleCategoryProductivity  = "productivity"
	ModuleCategoryCommunication = "communication"
	ModuleCategoryIntegration   = "integration"
)

// StringArray is a custom type for PostgreSQL text[] arrays
type StringArray []string

// Value implements the driver.Valuer interface for StringArray
func (s StringArray) Value() (driver.Value, error) {
	if s == nil {
		return nil, nil
	}
	return json.Marshal(s)
}

// Scan implements the sql.Scanner interface for StringArray
func (s *StringArray) Scan(value interface{}) error {
	if value == nil {
		*s = nil
		return nil
	}
	
	bytes, ok := value.([]byte)
	if !ok {
		return errors.New("failed to scan StringArray: invalid type")
	}
	
	// Try parsing as JSON array first (from GORM)
	if err := json.Unmarshal(bytes, s); err == nil {
		return nil
	}
	
	// Try parsing as PostgreSQL array format: {value1,value2}
	str := string(bytes)
	if len(str) >= 2 && str[0] == '{' && str[len(str)-1] == '}' {
		str = str[1 : len(str)-1]
		if str == "" {
			*s = []string{}
			return nil
		}
		// Simple split - doesn't handle escaped commas, but works for basic cases
		result := []string{}
		for _, item := range splitPgArray(str) {
			result = append(result, item)
		}
		*s = result
		return nil
	}
	
	return errors.New("failed to scan StringArray: unknown format")
}

// splitPgArray splits a PostgreSQL array string (without outer braces)
func splitPgArray(s string) []string {
	if s == "" {
		return []string{}
	}
	result := []string{}
	current := ""
	inQuotes := false
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c == '"' && (i == 0 || s[i-1] != '\\') {
			inQuotes = !inQuotes
		} else if c == ',' && !inQuotes {
			result = append(result, current)
			current = ""
		} else {
			current += string(c)
		}
	}
	if current != "" {
		result = append(result, current)
	}
	return result
}

// ModuleDefinition - Central registry of all available modules
// This is a reference table, typically seeded via migration
type ModuleDefinition struct {
	ID                   string         `gorm:"primaryKey" json:"id"` // e.g., 'pulse', 'review_workflow'
	Name                 string         `gorm:"not null" json:"name"`
	Description          string         `json:"description"`
	Icon                 string         `json:"icon"`
	Category             string         `gorm:"default:'core'" json:"category"`
	IsPremium            bool           `gorm:"default:false" json:"is_premium"`
	IsBeta               bool           `gorm:"default:false" json:"is_beta"`
	IsDeprecated         bool           `gorm:"default:false" json:"is_deprecated"`
	AvailableForHubTypes StringArray    `gorm:"type:text[]" json:"available_for_hub_types"`
	Dependencies         StringArray    `gorm:"type:text[]" json:"dependencies"`
	SettingsSchema       datatypes.JSON `gorm:"type:jsonb;default:'{}'" json:"settings_schema"`
	DisplayOrder         int            `gorm:"default:0" json:"display_order"`
	CreatedAt            time.Time      `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt            time.Time      `gorm:"autoUpdateTime" json:"updated_at"`
}

// TableName specifies the table name for ModuleDefinition
func (ModuleDefinition) TableName() string {
	return "module_definitions"
}

// HubModuleConfig - Per-table (hub) module enablement and configuration
type HubModuleConfig struct {
	ID        uuid.UUID      `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	TableID   uuid.UUID      `gorm:"type:uuid;not null;index" json:"table_id"`
	ModuleID  string         `gorm:"not null;index" json:"module_id"`
	IsEnabled bool           `gorm:"default:true" json:"is_enabled"`
	Settings  datatypes.JSON `gorm:"type:jsonb;default:'{}'" json:"settings"`
	EnabledBy *uuid.UUID     `gorm:"type:uuid" json:"enabled_by,omitempty"`
	EnabledAt time.Time      `gorm:"autoCreateTime" json:"enabled_at"`
	UpdatedAt time.Time      `gorm:"autoUpdateTime" json:"updated_at"`

	// Relationships
	Table  *Table            `gorm:"foreignKey:TableID" json:"table,omitempty"`
	Module *ModuleDefinition `gorm:"foreignKey:ModuleID" json:"module,omitempty"`
}

// TableName specifies the table name for HubModuleConfig
func (HubModuleConfig) TableName() string {
	return "hub_module_configs"
}

// BeforeCreate hook to generate UUID
func (h *HubModuleConfig) BeforeCreate(tx *gorm.DB) error {
	if h.ID == uuid.Nil {
		h.ID = uuid.New()
	}
	return nil
}

// TableWithModules extends Table with hub type and module configs
type TableWithModules struct {
	Table
	HubType       string            `gorm:"column:hub_type;default:'data'" json:"hub_type"`
	ModuleConfigs []HubModuleConfig `gorm:"foreignKey:TableID" json:"module_configs,omitempty"`
}

// TableName specifies the table name for TableWithModules
func (TableWithModules) TableName() string {
	return "data_tables"
}

// ============================================================
// DTO types for API responses
// ============================================================

// ModuleWithStatus combines module definition with its enabled status for a specific hub
type ModuleWithStatus struct {
	ModuleDefinition
	IsEnabled bool           `json:"is_enabled"`
	Settings  datatypes.JSON `json:"settings,omitempty"`
}

// HubModulesResponse is the API response for listing hub modules
type HubModulesResponse struct {
	TableID        uuid.UUID          `json:"table_id"`
	HubType        string             `json:"hub_type"`
	EnabledModules []ModuleWithStatus `json:"enabled_modules"`
	AvailableModules []ModuleWithStatus `json:"available_modules"`
}

// EnableModuleRequest is the request body for enabling a module on a hub
type EnableModuleRequest struct {
	ModuleID string         `json:"module_id" binding:"required"`
	Settings datatypes.JSON `json:"settings,omitempty"`
}

// UpdateHubTypeRequest is the request body for changing a table's hub type
type UpdateHubTypeRequest struct {
	HubType string `json:"hub_type" binding:"required,oneof=activities applications data"`
}

// ============================================================
// Helper methods
// ============================================================

// IsModuleAvailable checks if a module is available for this hub type
func (m *ModuleDefinition) IsModuleAvailable(hubType string) bool {
	for _, ht := range m.AvailableForHubTypes {
		if ht == hubType {
			return true
		}
	}
	return false
}

// HasDependency checks if a module has a specific dependency
func (m *ModuleDefinition) HasDependency(moduleID string) bool {
	for _, dep := range m.Dependencies {
		if dep == moduleID {
			return true
		}
	}
	return false
}

// GetEnabledModuleIDs returns a list of enabled module IDs for a hub
func GetEnabledModuleIDs(configs []HubModuleConfig) []string {
	var ids []string
	for _, config := range configs {
		if config.IsEnabled {
			ids = append(ids, config.ModuleID)
		}
	}
	return ids
}

// CanEnableModule checks if a module can be enabled (dependencies satisfied)
func CanEnableModule(module *ModuleDefinition, enabledModuleIDs []string) bool {
	// Check all dependencies are in the enabled list
	for _, dep := range module.Dependencies {
		found := false
		for _, enabled := range enabledModuleIDs {
			if enabled == dep {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	return true
}
